python/src/trezorlib/definitions.py

```diff
@@ -0,0 +1,145 @@
+DEFS_BASE_URL = "https://data.trezor.io/firmware/eth-definitions/"
+
+DEFINITIONS_DEV_SIGS_REQUIRED = 1
+DEFINITIONS_DEV_PUBLIC_KEYS = [
+    bytes.fromhex(key)
+    for key in ("db995fe25169d141cab9bbba92baa01f9f2e1ece7df4cb2ac05190f37fcc1f9d",)
+]
+
+DEFINITIONS_SIGS_REQUIRED = 2
+DEFINITIONS_PUBLIC_KEYS = [
+    bytes.fromhex(key)
+    for key in (
+        "4334996343623e462f0fc93311fef1484ca23d2ff1eec6df1fa8eb7e3573b3db",
+        "a9a22cc265a0cb1d6cb329bc0e60bc45df76b9ab28fb87b61136feaf8d8fdc96",
+        "b8d2b21de27124f0511f903ae7e60e07961810a0b8f28ea755fa50367a8a2b8b",
+    )
+]
+
+
+ProofFormat = c.PrefixedArray(c.Int8ul, c.Bytes(32))
+
+
+class DefinitionPayload(Struct):
+    magic: bytes
+    data_type: EthereumDefinitionType
+    timestamp: int
+    data: bytes
+
+    SUBCON = c.Struct(
+        "magic" / c.Const(FORMAT_MAGIC),
+        "data_type" / EnumAdapter(c.Int8ul, EthereumDefinitionType),
+        "timestamp" / c.Int32ul,
+        "data" / c.Prefixed(c.Int16ul, c.GreedyBytes),
+    )
+
+
+class Definition(Struct):
+    payload: DefinitionPayload = subcon(DefinitionPayload)
+    proof: t.List[bytes]
+    sigmask: int
+    signature: bytes
+
+    SUBCON = c.Struct(
+        "payload" / DefinitionPayload.SUBCON,
+        "proof" / ProofFormat,
+        "sigmask" / c.Int8ul,
+        "signature" / c.Bytes(64),
+    )
+
+    def verify(self, dev: bool = False) -> None:
+        payload = self.payload.build()
+        root = merkle_tree.evaluate_proof(payload, self.proof)
+        cosi.verify(
+            self.signature,
+            root,
+            DEFINITIONS_DEV_SIGS_REQUIRED,
+            DEFINITIONS_DEV_PUBLIC_KEYS,
+            self.sigmask,
+        )
+
+
+class Source:
+    def fetch_path(self, *components: str) -> t.Optional[bytes]:
+        raise NotImplementedError
+
+    def get_network_by_slip44(self, slip44: int) -> t.Optional[bytes]:
+        return self.fetch_path("slip44", str(slip44), "network.dat")
+
+    def get_network(self, chain_id: int) -> t.Optional[bytes]:
+        return self.fetch_path("chain-id", str(chain_id), "network.dat")
+
+    def get_token(self, chain_id: int, address: t.AnyStr) -> t.Optional[bytes]:
+        if isinstance(address, bytes):
+            address_str = address.hex()
+        elif address.startswith("0x"):
+            address_str = address[2:]
+        else:
+            address_str = address
+
+        address_str = address_str.lower()
+
+        return self.fetch_path("chain-id", f"{chain_id}", f"token-{address_str}.dat")
+
+
+class NullSource(Source):
+    def fetch_path(self, *components: str) -> t.Optional[bytes]:
+        return None
+
+
+class FilesystemSource(Source):
+    def __init__(self, root: Path) -> None:
+        self.root = root
+
+    def fetch_path(self, *components: str) -> t.Optional[bytes]:
+        path = self.root.joinpath(*components)
+        if not path.exists():
+            LOG.info("Requested definition at %s was not found", str(path))
```

Comment:
`...was not found` messages could be `LOG.error` or `LOG.warning`

Also, why not using f-strings?

Reply:
> ...was not found messages could be LOG.error or LOG.warning

Not for a missing token, esp. with how we opportunistically try to load a definition any time any sort of data (not just erc20 transaction) is being sent.
Warning for a missing network is debatable, but the way we use the definitions as strictly optional, it just seems unnecessary. (e.g., if we decide to remove the builtins from generated data, you would get a warning for every builtin).

> Also, why not using f-strings?

it's a `logging` thing. The format string and the arguments are actually stored separately and interpolated only when someone asks to output the log. This mechanism only supports `%` formatting.
In here it wouldn't really matter, but I went with how "things are done" with logging.

For that matter, stringifying the path is useless here because `%s` will do that.

Reply:
removed str(path) in rebase